rust/tools/authorization-logic/README.md

```diff
@@ -0,0 +1,209 @@
+At time of writing, this protoype is about the same as SecPal. A longer 
+description of authorization logic is given below, followed by instructions.
+
+This authoriation logic compiles programs from the source language into the 
+[Souffle](https://souffle-lang.github.io/index.html) implementation of datalog.
+Souffle should be installed by following 
+[these instructions](https://souffle-lang.github.io/install).
+
+## Summary of Authorization Logic
+
+### Logic Programming
+Authorization logics extend some logic programming language (e.g. Datalog, 
+Prolog), and permit an action only when a specific predicate can be proved 
+using inference rules and facts written in this language. Generally, 
+authorization logic is paired with a reference monitor that interprets the 
+policy language and permits the guarded actions only when presented with a 
+proof of the proposition used to grant access.
+
+### Principals and decentralization
+The key feature of authorization logic is that facts are decentralized - each 
+fact or rule is attributed to a principal (or actor). For example, HR says 
+isManager(Alice) means that HR believes or claims that isManager(Alice) is 
+true. Decentralizing facts in this way ensures that a resource owner retains 
+control over access to resources in an environment with mutually distrusting 
+principals. As an example, to use the rule
+```
+"FileServer" says canAccess(file, requestor) :- isManager(requestor),                
+                                              isValidFile(file).
+```
+requires proving that FileServer says isManager(requestor) and FileServer says 
+isValidFile(file). 
+
+### Delegation
+Because authorization logic models beliefs held by various parties rather than 
+a single global truth believed by all, we need a way to export facts from one 
+principal to another. In authorization logic, this is done by delegation. For 
+example:
+
+```
+FileServer says HR canSay isManager(x).
+```
+means that HR canSay isManager(x) implies FileServer says isManager(x). The 
+exact language-level construct for delegation has varied among authorization 
+logics, but the key point is that this gives one principal (FileServer) a way 
+of specifying when it believes claims made by another (HR). Delegations can 
+also be done conditionally by placing them on the left of a proof rule.
+
+
+### Authenticating policies
+When FileServer imports a policy statement from HR, we want to know that a 
+claim genuinely originated from HR. To authenticate policy statements, each 
+principal is also associated with an asymmetric key-pair. When a policy 
+statement is exported by a principal, the statement is digitally signed, the 
+signature is sent to the recipient along with a serialization of the policy 
+claim, and the recipient uses the public key to check the signature.
+
+## Instructions
+
+### Running a program
+To compile a program, run:
+```
+cargo run FILENAME INPUT_DIR OUTPUT_DIR
```

Comment:
At the moment, only a single file can be compiled at a time. (A program can still involve multiple files via importing/exporting which involves checking signatures as described in Importing/Exporting statements below).

A program called FILENAME stored in INPUT_DIR is compiled producing OUTPUT_DIR/FILENAME.dl. I am writing the text to be clearer locally.

(I realise also that the way this actually works is not very standard but this is easy to change later and because this is a prototype, so I priorotised just getting it working somehow).